solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/PrefetchingFieldValueFeature.java

```diff
@@ -0,0 +1,149 @@
+          ": field must be provided");
+    }
+  }
+
+  public PrefetchingFieldValueFeature(String name, Map<String,Object> params) {
+    super(name, params);
+  }
+
+  @Override
+  public FeatureWeight createWeight(IndexSearcher searcher, boolean needsScores,
+      SolrQueryRequest request, Query originalQuery, Map<String,String[]> efi)
+          throws IOException {
+    return new PrefetchingFieldValueFeatureWeight(searcher, request, originalQuery, efi);
+  }
+
+  @VisibleForTesting
+  public Set<String> getPrefetchFields(){
+    return prefetchFields;
+  }
+
+  public class PrefetchingFieldValueFeatureWeight extends FieldValueFeatureWeight {
+    private final SchemaField schemaField;
+    private final SolrDocumentFetcher docFetcher;
+
+    public PrefetchingFieldValueFeatureWeight(IndexSearcher searcher,
+        SolrQueryRequest request, Query originalQuery, Map<String,String[]> efi) {
+      super(searcher, request, originalQuery, efi);
+      if (searcher instanceof SolrIndexSearcher) {
+        schemaField = ((SolrIndexSearcher) searcher).getSchema().getFieldOrNull(field);
+      } else {
+        schemaField = null;
+      }
+      this.docFetcher = request.getSearcher().getDocFetcher();
+    }
+
+    /**
+     * Return a FeatureScorer that work with stored fields and makes use of the cache if the configured field is stored
+     * and has no docValues.
+     * Otherwise, delegate the work to the FieldValueFeature.
+     *
+     * @param context the segment this FeatureScorer is working with
+     * @return FeatureScorer for the current segment and field
+     * @throws IOException as defined by abstract class Feature
+     */
+    @Override
+    public FeatureScorer scorer(LeafReaderContext context) throws IOException {
+      if (schemaField != null && !schemaField.stored() && schemaField.hasDocValues()) {
+        return super.scorer(context);
+      }
+      return new PrefetchingFieldValueFeatureScorer(this, context,
+          DocIdSetIterator.all(DocIdSetIterator.NO_MORE_DOCS));
+    }
+
+    /**
+     * A FeatureScorer that reads the stored value for a field
+     * docFetcher does not request a single field but all the prefetchFields to improve performance through caching
+     */
+    public class PrefetchingFieldValueFeatureScorer extends FieldValueFeatureScorer {
+
+      public PrefetchingFieldValueFeatureScorer(FeatureWeight weight,
+          LeafReaderContext context, DocIdSetIterator itr) {
+        super(weight, context, itr);
+      }
+
+      @Override
+      public float score() throws IOException {
+        try {
+          final Document document = docFetcher.doc(context.docBase + itr.docID(), prefetchFields);
+
+          return super.parseStoredFieldValue(document.getField(field));
+        } catch (final IOException e) {
+          throw new FeatureException(
+              e.toString() + ": " +
+                  "Unable to extract feature for "
+                  + name, e);
```

Comment:
> overriding toString(String) or toString() might be something to consider

I am wondering whether it could also be an option to add the prefetchFields to `paramsToMap()`. I did not look into the side effects / benefits this might have yet.

Reply:
> Also with troubleshooting in mind, I wonder if `prefetchFields` being not just a set but an ordered set might be useful (_assuming_ that ordering would _not_ have a runtime performance cost).

> I assume that your intention behind the ordered set is, that if the user wanted to prefetch fields `A, B, C, D` and an exception regarding field `C` would happen, they would know that `A` and `B` worked. Did I understand you correctly?

I think my thoughts were a bit vague on that actually, thank you for encouraging me to organise them a bit more :)

* Depending on how folks name their features and how many features there are, if one had to interpret error/exception logging with the details then `fA, fC, ... fX, fZ` i.e. alphabetical is (subjectively) more human friendly than some random ordering that might arise via the HashSet implementation.

* Given the same inputs in the same order, will two Solr nodes come up with the same HashSet ordering for them or not? What if the nodes run different Java versions or one node has been running a long time and another is recently started? I don't know the answer to those questions but if our two hypethical nodes both complained about some features but one complained about `fZ, fA, fX, fB` and the other about `fA, fZ, fB, fX` then it might be tricky to interpret that they are both complaining about the same thing. Or if there is a difference between them then sorted ordering should make it easier to spot what the difference is.

* In terms of knowing that a field C exception means that fields A and B still worked, no, that's actually not something that had occurred to me. But now that you mention it in that way i.e. the potential for one feature or one field to affect some but not all of the other features ... I wonder if the exception handling could log (say) a WARNing and then do a noPrefetching retry, something like this:

```
  try {
    document = docFetcher.doc(context.docBase + itr.docID(), prefetchFields);
    ...
  } catch {
    ... log warning ...
    try {
      document = docFetcher.doc(context.docBase + itr.docID(), fieldAsSet);
      ... log info to say it worked ...
      ...
    } catch {
      throw exception
    }
  }
```
(And I note that such a fallback-to-no-prefetching mechanism would reduce the need for a `disablePrefetching` flag on the request level i.e. the system would "self debug" i.e. the unproblematic fields/features would log a warning and then an info to indicate that the fallback worked and problematic fields/features would log a warning and then presumably would fail again and throw an exception after the single-field retry.)

Reply:
Thank you for the pointers. :)
After trying `paramsToMap()` and `toString()` I opted for overriding `paramsToMap()` because I think it is very useful to see the prefetch fields when performing the get request on the feature store. ( fa1fb82 )
I had to do an adjustment in `paramsToMap()`: If there are no prefetch fields then we have to return an empty set instead of null to prevent a NPE further down the line.

While adding tests for `paramsToMap()` I realized, that the `doTestParamsToMap()` test did not work.
Two `PrefetchingFieldValueFeatures` are not equal, if one of them is initialized with a params map that is missing the prefetch fields and the other one is initialized with am empty set for the prefetchFields. (Because their params differ and therefore one has prefetchFields = null, the other one has an empty set.)

However, the prefetchFields should never be empty / null, because `featuresAsManagedResources()` is called after  `preparePrefetchingFieldValueFeatures` where the prefetchFields are initialized.
So I would opt to ignore this.
If you disagree on ignoring that we could fix it by a) overriding `equals()` (quite a lot of code) or b) always initializing prefetchFields with an empty set.

I had to add another setter for the prefetchFields because when the params are loaded from the storage ( https://github.com/apache/lucene-solr/blob/a92a05e195b775b30ca410bc0a26e8e79e7b3bfb/solr/core/src/java/org/apache/solr/rest/ManagedResourceStorage.java#L489 ) the information that the prefetchFields are a Set is lost.

Reply:
Hi @tomglk! I really like your idea of including the prefetchFields in the stored representation that one gets when downloading from the feature store or when inspecting in ZooKeeper.

In https://github.com/cpoerschke/solr/commit/df2864a95e1a83ee446781042fe119981551a932 I explored splitting PrefetchingFieldValueFeature.prefetchFields into three (configured, computed, effective) and 'self' not being part of the params-to-map prefetch fields. But I am not yet 100% convinced by that splitting approach and the commit also includes some small others bits mixed in. What do you think?

On your (not) overriding `equals` point, that's a good one. I think it's okay to map `null` to `empty set` in `paramsToMap` like you did.

Reply:
I talked about your commit in a comment below ( https://github.com/apache/solr/pull/166#discussion_r650527430 and following ), so I'm just leaving the pointer here. :)